quora_dataset.py

```diff
@@ -63,78 +62,64 @@ def get_examples(self):
+        for e in self._examples:
             # each positive example is considered an edge
             uf.union(e.qid1, e.qid2)
 
-        #  disjoint_set is a map of set name to its members
+        # disjoint_set is a map of set name to its members
         disjoint_set = defaultdict(list)
         for qid in pos_qids:
             set_name = uf.find(qid)
             disjoint_set[set_name].append(qid)
 
         # construct training pairs
-        set_names = list(disjoint_set.keys())
-        random.shuffle(set_names)
+        items = list(disjoint_set.items())
+        random.shuffle(items)
+        disjoint_set = dict(items)
+
         train_qid_pairs = set()
-        for sn in set_names:
-            dset = disjoint_set[sn]
+        for key in list(disjoint_set.keys()):
+            dset = disjoint_set.pop(key)
             pairs = itertools.combinations(dset, 2)
             train_qid_pairs.update(pairs)
 
             # might go over the target train_size a bit
             if len(train_qid_pairs) >= train_size:
                 break
-
-        # filter out examples whose qids are already in the train_qid_pairs
-        test_examples = []
         train_qids = set(flatmap(train_qid_pairs))
-        for e in self._examples:
-            if not e.is_duplicate:
-                test_examples.append(e)
-            else:
-                if e.qid1 not in train_qids and e.qid2 not in train_qids:
-                    test_examples.append(e)
 
         # build positive test_qid_pairs
         test_qid_pairs = set()
-        random.shuffle(test_examples)
         candidate_ids = set()
         query2result_list = []
-        for e in test_examples:
-            if e.qid1 in candidate_ids or e.qid2 in candidate_ids:
-                continue
-
-            if e.is_duplicate:
-                # doesn't matter which of qid1 or qid2 to use here.  they are in the same set
-                set_members = disjoint_set[uf.find(e.qid1)]
-                pairs = list(itertools.combinations(set_members, 2))
-                test_qid_pairs.update(pairs)
-
-                # set members are qids which are relevant to each other
-                # every member can be considered a test query
-                candidate_ids.update(set_members)
-                for qid in set_members:
-                    result = set_members.copy()
-                    # remove query from result
-                    result.remove(qid)
-                    query2result_list.append((qid, result))
-
-            candidate_ids.update([e.qid1, e.qid2])
-            if len(query2result_list) >= retrieval_size:
+        for set_members in list(disjoint_set.values()):
+            # set members are qids which are relevant to each other
+            # every member can be considered a test query
+
+            # Store limited number of queries for the purposes of retrieval (but all relevant results)
+            candidate_ids.update(set_members)
+            for qid in set_members[:1]:
```

Comment:
We are just taking the first set member as the query.  Is this intentional?

Reply:
It is, if we store all then the number of test queries and the number of candidates will be the same. The logic here could be changed so that we get closer to `2.55 candidates/query`.